imap_processing/ccsds/excel_to_xtce.py

```diff
@@ -0,0 +1,419 @@
+"""
+Convert an Excel file of packet definitions into the XTCE format.
+
+This script reads in an Excel file containing packet definitions and converts
+them into an XTCE file.
+
+.. code::
+  imap_xtce /path/to/excel_file.xlsx --output /path/to/output.xml
+"""
+
+import argparse
+import xml.etree.ElementTree as Et
+from importlib.util import find_spec
+from pathlib import Path
+
+import pandas as pd
+
+_CCSDS_PARAMETERS = [
+    {
+        "name": "VERSION",
+        "lengthInBits": 3,
+        "description": "CCSDS Packet Version Number (always 0)",
+    },
+    {
+        "name": "TYPE",
+        "lengthInBits": 1,
+        "description": "CCSDS Packet Type Indicator (0=telemetry)",
+    },
+    {
+        "name": "SEC_HDR_FLG",
+        "lengthInBits": 1,
+        "description": "CCSDS Packet Secondary Header Flag (always 1)",
+    },
+    {
+        "name": "PKT_APID",
+        "lengthInBits": 11,
+        "description": "CCSDS Packet Application Process ID",
+    },
+    {
+        "name": "SEQ_FLGS",
+        "lengthInBits": 2,
+        "description": "CCSDS Packet Grouping Flags (3=not part of group)",
+    },
+    {
+        "name": "SRC_SEQ_CTR",
+        "lengthInBits": 14,
+        "description": "CCSDS Packet Sequence Count "
+        "(increments with each new packet)",
+    },
+    {
+        "name": "PKT_LEN",
+        "lengthInBits": 16,
+        "description": "CCSDS Packet Length "
+        "(number of bytes after Packet length minus 1)",
+    },
+]
+
+
+class XTCEGenerator:
+    """
+    Automatically generate XTCE files from excel definition files.
+
+    The excel file should have the following columns: mnemonic, sequence, lengthInBits,
+    startBit, dataType, convertAs, units, source, and either shortDescription or
+    longDescription.
+
+    This class will correctly generate a CCSDS header and the provided data types.
+
+    It is intended for use as a first pass of XTCE generation, and most cases, the
+    packet definitions will require manual updates.
+
+    Use ``to_xml`` to write the output xml file.
+
+    Parameters
+    ----------
+    path_to_excel_file : Path
+        Path to the excel file.
+    """
+
+    def __init__(self, path_to_excel_file: Path):
+        # Read in all sheets from the excel file
+        self.sheets = pd.read_excel(path_to_excel_file, sheet_name=None)
+        # Set up the packet mapping from packetName to Apid
+        packet_sheet = self.sheets["Packets"]
+        self._packet_mapping = packet_sheet.set_index("packetName")["apId"].to_dict()
```

Comment:
I tried with SWE packet definition and it seems to use `apIdHex` instead of `apId`.

Reply:
Another thing I was wondering is if we can add option to filter which packets we want to write to xml? Eg. write xtce definition for only science packets and leave out non-science ones.

Reply:
> I tried this with SWE packet definition and SWE seems to use apIdHex instead of apId. They are same information but saved in different ways.

Can you send me their Excel file? Do you mean they don't have an apid column? Or that they have the hex values in the apid column? We could probably have a fallback here to calculate the apid from the hex if needed.

> Another thing I was wondering is if we can add option to filter which packets we want to write to xml? Eg. write xtce definition for only science packets and leave out non-science ones. Some way to take in user input for that.

I debated this and wasn't sure whether that would be useful or not. My thought was that it might be easier to just delete entries from the Excel file first if you don't want to include those rather than having to learn a new input/output flag to use? For example, Hi has a lot of packets listed, but there are only a few tabs listed so then the program outputs:
```
Packet definition for H45_BOOT_HK not found in the excel file.
Packet definition for H45_BOOT_MEMDMP not found in the excel file.
```
So if you don't want certain entries in the XTCE then you could just delete those tabs from the spreadsheet.

Reply:
Deleting tabs/packets we don't want include works for me!